code/go/0chain.net/smartcontract/storagesc/stakepool.go

```diff
@@ -89,32 +90,52 @@ func (sp *stakePool) Decode(input []byte) error {
 }
 
 // save the stake pool
-func (sp *stakePool) save(sscKey, blobberID string,
+func (sp *stakePool) save(providerType spenum.Provider, providerID string,
 	balances chainstate.StateContextI) (err error) {
-
-	r, err := balances.InsertTrieNode(stakePoolKey(sscKey, blobberID), sp)
+	r, err := balances.InsertTrieNode(stakePoolKey(providerType, providerID), sp)
 	logging.Logger.Debug("after stake pool save", zap.String("root", r))
 
-	data := dbs.DbUpdates{
-		Id: blobberID,
-		Updates: map[string]interface{}{
-			"offers_total": int64(sp.TotalOffers),
-		},
-	}
-	balances.EmitEvent(event.TypeStats, event.TagUpdateBlobber, blobberID, data)
+	sp.emitSaveEvent(providerType, providerID, balances)
 
 	return
 }
 
+func (sp *stakePool) emitSaveEvent(providerType spenum.Provider, providerID string, balances chainstate.StateContextI) {
+	data := dbs.DbUpdates{
+		Id: providerID,
+		Updates: map[string]interface{}{},
+	}
+	switch providerType {
+	case spenum.Blobber:
+		data.Updates["offers_total"] = int64(sp.TotalOffers)
+		balances.EmitEvent(event.TypeStats, event.TagUpdateBlobber, providerID, data)
+	case spenum.Validator:
+		//todo: emit validator event
+	default:
+		logging.Logger.Error("invalid providerType in stakepool SaveEvent")
+	}
+}
+
 // The cleanStake() is stake amount without delegate pools want to unstake.
 func (sp *stakePool) cleanStake() (stake currency.Coin, err error) {
 	staked, err := sp.stake()
 	if err != nil {
 		return 0, err
 	}
+
 	return staked - sp.TotalUnStake, nil
 }
 
+func (sp *stakePool) stakeByProvider(providerType spenum.Provider, providerID string, balances chainstate.StateContextI) error{
```

Comment:
it should be emit stake event?

Reply:
No, the primary goal is to perform the staking operation, emitting event is secondary. The event is packaged here so that we don't have to manually call the emit event method every time this staking is done.